src/shared/collectErrors/CollectErrors.ts

```diff
@@ -0,0 +1,15 @@
+import { ParseError } from '@shared/ParseError'
+
+type Validator<T> = (val: T) => ParseError | null
+
+export const collectErrors = <T>(
```

Comment:
Since this is collecting parsing errors, maybe we could be more explicit

```suggestion
export const collectParseErrors = <T>(
```